### PR TITLE
Revise Terms

### PR DESCRIPTION
**Pull request by AlexTugarev**
#### How to test
1. TC – login **and** signup should work as usual
2. TC – run sql `update d_b_terms_acceptance_entry set termsRevision='2020-10';` in order to test an update. 
  (feel free to ask me to run it for you 🙏🏻 )
  then:
 2.1. `/workspaces` page must not be affected, neither any running workspace
 2.2. starting new/existing workspaces must be blocked by `We've update ToS` page
 2.3. re-login should be blocked by `We've update ToS` page

### CODE REVIEW

File: .gitpod.yml [hunk 1]
```diff
@@ -34,6 +34,7 @@ tasks:
   - name: Go
     init: leeway exec --filter-type go -v -- go get -v ./...
     openMode: split-right
+privileged: true
```
Comment (csweichel) [inline .gitpod.yml:37]: We no longer support this in workspace config.
If you want to use this feature flag, please set it on your user.

**Comment by akosyakov** [comment]
I logged in, updated db, reload the page and got such state:
Screenshot 2020-12-01 at 15 06 17


i.e. I see my workspaces and login button too.

**Comment by AlexTugarev** [comment]
@akosyakov thanks for the test!

It went broken by accident in the commit after our session :-( 
(https://github.com/gitpod-io/gitpod/compare/ba361ecbba54c56c96aa20ed7c03d755c507f162..59fea49a454a797f1bfd797050826131f8ff3c3d)

**Comment by csweichel** [comment]
When signing up, I found the sentence
> Read the terms online.

a bit odd. Can I read the terms offline?

**Comment by csweichel** [comment]
Just ran the test and it works like a charm.

When asked to accept the updated ToS I clicked decline, and was logged out which felt a bit sudden/unexpected. Maybe a message that that's gonna happen would be nice @gtsiolis?

Once I logged back in I was asked to accept the updated ToS right away, i.e. I could not get to the workspace list without accepting the updated ToS. While this makes perfect sense, it feels a bit inconsistent. I can still use the workspace list without accepting the new ToS, e.g. stop workspaces or download their content - as long as I don't log out. Once I log out, that's gone, too. I don't think this is a big issue at all - and if we wanted to do something about it (that's a big if at this point), we could show yet another message in the dashboard to inform users that the ToS have changed and that they'll have to accept them eventually.

**Comment by csweichel** [comment]
Actually, without accepting the new ToS, I can still click `download workspace`, but see a 404:
image

On gitpod-staging.com the workspace download works, so it's not broken in general/on master.

In the traces there's nothing (we're not tracing workspace download it seems), and in the server logs things look fine:
```
{"component":"server","severity":"INFO","time":"2020-12-02T08:05:33.628Z","environment":"devstaging","region":"europe-west1","context":{"workspaceId":"f7bc87fd-93d5-4d3b-88e6-f7ff6a200922","userId":"95d9fc9e-e79d-48f4-9574-6a7ee615ffa7"},"message":"user is downloading workspace content"}
```
This log message indicates that the only [404-returning branch](https://github.com/gitpod-io/gitpod/blob/653b48ae5c64fb01bd01800f0a1768c49844a121/components/server/src/workspace/workspace-download-service.ts#L40-L44) in downloadWorkspace is not the culprit.

The page 404's before we try and talk to the server using websockets.

**Comment by akosyakov** [comment]
I got in such state:
Screenshot 2020-12-02 at 09 16 24

- Start a workspace, then log out in another tab, reload the workspace page.

**Comment by AlexTugarev** [comment]
@csweichel only create/start workspaces are guarded, i.e. if downloading workspace is broken, this is a separate issue.

**Comment by csweichel** [comment]
> @csweichel only create/start workspaces are guarded, i.e. if downloading workspace is broken, this is a separate issue.

makes sense. Upon closer inspection it turns out that it's because we're using minio in dev-staging.

Sorry for the noise :(

**Comment by akosyakov** [comment]
Generally works well, but I'm getting ToS for running workspaces, i.e. reload the page after changing the DB.

File: components/dashboard/src/components/start-workspace.tsx [hunk 2]
```diff
@@ -322,16 +358,50 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
...
+        }
+        if (error?.code === ErrorCodes.NOT_AUTHENTICATED) {
+            if (error?.data.host && error?.data.scopes && error?.data.messageHint) {
+                return (
+                    <ApplicationFrame service={this.props.service}>
+                        <ShowUnauthorizedError data={error?.data} />
+                    </ApplicationFrame>
+                );
+            }
+            console.log(error?.data);
```
Comment (akosyakov) [inline components/dashboard/src/components/start-workspace.tsx:384]: leftover

File: components/dashboard/src/components/start-workspace.tsx [hunk 3]
```diff
@@ -150,23 +160,48 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
...
+                    case ErrorCodes.SETUP_REQUIRED:
+                        this.redirectTo(new GitpodHostUrl(window.location.toString()).with({ pathname: "first-steps" }).toString());
+                        return;
+                    case ErrorCodes.USER_TERMS_ACCEPTANCE_REQUIRED:
+                        const thisUrl = window.location.toString();
+                        this.redirectTo(new GitpodHostUrl(thisUrl).withApi({ pathname: "/tos", search: `returnTo=${encodeURIComponent(thisUrl)}` }).toString());
+                        return;
+                    case ErrorCodes.NOT_AUTHENTICATED:
+                        if (err.data) {
+                            this.setState({ error: { message: err.toString(), data: err.data, code: err.code } });
```
Comment (akosyakov) [inline components/dashboard/src/components/start-workspace.tsx:184]: err.message

File: components/dashboard/src/components/start-workspace.tsx [hunk 4]
```diff
@@ -150,23 +160,48 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
                 }
             })
             .catch(err => {
-                this.setErrorState(err, defaultErrMessage);
+                this.setErrorState(err);
```
Comment (akosyakov) [inline components/dashboard/src/components/start-workspace.tsx:163]: defaultErrMessage is lost for err.message is empty

**Comment by gtsiolis** [comment]
> When asked to accept the updated ToS I clicked decline, and was logged out which felt a bit sudden/unexpected. Maybe a message that that's gonna happen would be nice @gtsiolis?

@csweichel you're right! This is the reason why we've updated this button label to be more descriptive in https://github.com/gitpod-com/gitpod/pull/4961/files#diff-ecd398c39dfd28807e8affbd7102375257fbdeb014c3d0007be2b65c7432c626R178.

We should probably update this to be the same in this repository, right? /cc @AlexTugarev

**Review by akosyakov — APPROVED** [review]
LGTM

**Comment by geropl** [comment]
Tested successfully, starting with code review...

File: components/dashboard/src/components/start-workspace.tsx [hunk 5]
```diff
@@ -158,15 +164,43 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
     protected setErrorState(err: any, defaultErrorMessage?: string) {
         let errorMessage = defaultErrorMessage;
         let errorCode = undefined;
+        let errorData = undefined;
         if (err instanceof ResponseError) {
             errorCode = err.code;
+            errorData = err.data;
+            errorMessage = err.message;
+
+            switch (err.code) {
```
Comment (geropl) [inline components/dashboard/src/components/start-workspace.tsx:173]: Why not move the handling to "renderError"?

File: components/dashboard/src/components/start-workspace.tsx [hunk 6]
```diff
@@ -322,16 +356,49 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
...
+        if (errorCode === ErrorCodes.NOT_AUTHENTICATED) {
+            if (errorData?.host && errorData?.scopes && errorData?.messageHint) {
+                return (
+                    <ApplicationFrame service={this.props.service}>
+                        <ShowUnauthorizedError data={errorData} />
+                    </ApplicationFrame>
+                );
+            }
+        }
+        if (errorCode === ErrorCodes.PLAN_DOES_NOT_ALLOW_PRIVATE_REPOS) {
```
Comment (geropl) [inline components/dashboard/src/components/start-workspace.tsx:383]: This is an IO concern which is handled in the `startErrorRenderer`.

File: components/gitpod-db/src/typeorm/entity/db-terms-acceptance-entry.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { Column, Entity, PrimaryColumn } from "typeorm";
+import { TermsAcceptanceEntry } from "@gitpod/gitpod-protocol";
+
+@Entity()
```
Comment (geropl) [inline components/gitpod-db/src/typeorm/entity/db-terms-acceptance-entry.ts:10]: A short comment about how the DB is used would be nice. E.g.: "single entry per user which is overriden/updated over time"

File: components/gitpod-db/src/typeorm/migration/1606163573103-TermsAcceptance.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class TermsAcceptance1606163573103 implements MigrationInterface {
+
+    public async up(queryRunner: QueryRunner): Promise<any> {
+        await queryRunner.query("CREATE TABLE IF NOT EXISTS `d_b_terms_acceptance_entry` ( `userId` char(36) NOT NULL, `termsRevision` varchar(255) NOT NULL, `acceptionTime` varchar(255) NOT NULL, PRIMARY KEY (userId) ) ENGINE=InnoDB");
```
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1606163573103-TermsAcceptance.ts:12]: Using `tableExists` from `./helper/helper.ts` is for stable.

**Comment by geropl** [comment]
General question: Is Tos meant to be EE code? If yes, we should move the error handling in the UI + the `TermsProvider` etc. there.

**Comment by geropl** [comment]
Done! :+1:

**Review by meysholdt — COMMENTED** [review]
comment

**Review by meysholdt — APPROVED** [review]
LGTM

File: components/dashboard/src/components/start-workspace.tsx (hunk 5 above)
Comment (AlexTugarev) [reply components/dashboard/src/components/start-workspace.tsx:173]: 1. (and foremost) I tried to align with `create-workspace.tsx` as much as possible and avoided adding unknown bits
2. here, it's mostly about redirecting right after getting an error, not so much about rendering error state.

File: components/dashboard/src/components/start-workspace.tsx (hunk 6 above)
Comment (AlexTugarev) [reply components/dashboard/src/components/start-workspace.tsx:383]: removing...

**Comment by AlexTugarev** [comment]
> General question: Is Tos meant to be EE code? If yes, we should move the error handling in the UI + the TermsProvider etc. there.

@geropl, terms' already there and it's not EE code. Why considering a change? 🤷🏻‍♂️

**Comment by geropl** [comment]
> , terms' already there and it's not EE code. Why considering a change? 🤷🏻‍♂️

@AlexTugarev I had the impression from past discussions; just wanted to verify.

File: components/dashboard/src/components/start-workspace.tsx (hunk 5 above)
Comment (geropl) [reply components/dashboard/src/components/start-workspace.tsx:173]: > (and foremost) I tried to align with create-workspace.tsx as much as possible and avoided adding unknown bits

**Review by geropl — APPROVED** [review]
:+1:

**Pull request merged**